### PR TITLE
Add support for the SARIF format and ruleId lookup, for the Violations metric

### LINKED ISSUE #6314: Add support for the Sarif format in combination with the Violations metric
**Describe the solution you'd like**
Add support for the Sarif format in combination with the Violations metric. Specific for this usecase to support the output of Robocop: https://robocop.readthedocs.io/en/stable/ a tool that performs static code analysis of Robot Framework code.

**Tasks**

- Add sarif as source for the violations metric in the data model.
- Support looking up the rules by using the `ruleId` in addition to used the `ruleIndex` which we already support. 

**Additional context**
Example .sarif.json:

```json
{
    "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
    "version": "2.1.0",
    "runs": [
        {
            "tool": {
                "driver": {
                    "name": "Robocop",
                    "semanticVersion": "3.1.1",
                    "informationUri": "https://robocop.readthedocs.io/",
                    "rules": [
                        {
                            "id": "0201",
                            "name": "missing-doc-keyword",
                            "helpUri": "https://robocop.readthedocs.io/en/stable/rules.html#missing-doc-keyword",
                            "shortDescription": {
                                "text": "Missing documentation in '{{ name }}' keyword"
                            },
                            "fullDescription": {
                                "text": "\nYou can add documentation to keyword using following syntax::\n\n    Keyword\n        [Documentation]  Keyword documentation\n        Keyword Step\n        Other Step\n\n"
                            },
                            "defaultConfiguration": {
                                "level": "warning"
                            },
                            "help": {
                                "text": "\nYou can add documentation to keyword using following syntax::\n\n    Keyword\n        [Documentation]  Keyword documentation\n        Keyword Step\n        Other Step\n\n",
                                "markdown": "\nYou can add documentation to keyword using following syntax::\n\n    Keyword\n        [Documentation]  Keyword documentation\n        Keyword Step\n        Other Step\n\n"
                            }
                        },
                        {
                            "id": "0202",
                            "name": "missing-doc-test-case",
                            "helpUri": "https://robocop.readthedocs.io/en/stable/rules.html#missing-doc-test-case",
                            "shortDescription": {
                                "text": "Missing documentation in '{{ name }}' test case"
                            },
                            "fullDescription": {
                                "text": "\nYou can add documentation to test case using following syntax::\n\n    Test\n        [Documentation]  Test documentation\n        Keyword Step\n        Other Step\n\nThe rule by default ignores templated test cases but it can be configured with::\n\n    robocop --configure missing-doc-test-case:ignore_templated:False\n\nPossible values are: Yes / 1 / True (default) or No / False / 0.\n"
                            },
                            "defaultConfiguration": {
                                "level": "warning"
                            },
                            "help": {
                                "text": "\nYou can add documentation to test case using following syntax::\n\n    Test\n        [Documentation]  Test documentation\n        Keyword Step\n        Other Step\n\nThe rule by default ignores templated test cases but it can be configured with::\n\n    robocop --configure missing-doc-test-case:ignore_templated:False\n\nPossible values are: Yes / 1 / True (default) or No / False / 0.\n",
                                "markdown": "\nYou can add documentation to test case using following syntax::\n\n    Test\n        [Documentation]  Test documentation\n        Keyword Step\n        Other Step\n\nThe rule by default ignores templated test cases but it can be configured with::\n\n    robocop --configure missing-doc-test-case:ignore_templated:False\n\nPossible values are: Yes / 1 / True (default) or No / False / 0.\n"
                            }
                        }
                    ]
                }
            },
            "automationDetails": {
                "id": "robocop/"
            },
            "results": [
                {
                    "ruleId": "0302",
                    "level": "warning",
                    "message": {
                        "text": "Keyword name 'XYZ' does not follow case convention"
                    },
                    "locations": [
                        {
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "testcase1.robot",
                                    "uriBaseId": "%SRCROOT%"
                                },
                                "region": {
                                    "startLine": 7,
                                    "endLine": 7,
                                    "startColumn": 19,
                                    "endColumn": 50
                                }
                            }
                        }
                    ]
                },
                {
                    "ruleId": "0803",
                    "level": "error",
                    "message": {
                        "text": "Multiple variables with name 'XYZ' in Variables section (first occurrence in line 53). Note that Robot Framework is case-insensitive"
                    },
                    "locations": [
                        {
                            "physicalLocation": {
                                "artifactLocation": {
                                    "uri": "testcase2.robot",
                                    "uriBaseId": "%SRCROOT%"
                                },
                                "region": {
                                    "startLine": 54,
                                    "endLine": 54,
                                    "startColumn": 1,
                                    "endColumn": 15
                                }
                            }
                        }
                    ]
                }
            ]
        }
    ]
}
```

### PR DESCRIPTION
**Pull request by wkoot**
Closes #6314

### CODE REVIEW

File: components/collector/src/source_collectors/sarif/base.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+from typing import cast
+
+from base_collectors import JSONFileSourceCollector
+from collector_utilities.type import JSON, JSONDict
+from model import Entities, Entity
+
+
+class SARIFJSONBase(JSONFileSourceCollector):
+    """SARIF collector for security warnings."""
```
Comment (fniessink) [inline components/collector/src/source_collectors/sarif/base.py]: No longer just security warnings, right?

```suggestion
    """Base class for collectors that read SARIF JSON files."""
```

File: components/collector/src/source_collectors/sarif/base.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    def _parse_json(self, json: JSON, filename: str) -> Entities:
+        """Override to parse the security warnings from the JSON."""
```
Comment (fniessink) [inline components/collector/src/source_collectors/sarif/base.py]:
```suggestion
        """Override to parse the analysis results from the SARIF JSON."""
```

File: components/collector/src/source_collectors/sarif/base.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        entities = Entities()
+        runs = cast(JSONDict, json).get("runs", [])
+        for run in runs:
+            rules = run["tool"]["driver"]["rules"]
+            for result in run["results"]:
+                rule = self._lookup_result_rule(result, rules)
```
Comment (fniessink) [inline components/collector/src/source_collectors/sarif/base.py]:
```suggestion
                rule = self._lookup_violated_rule(result, rules)
```

File: components/collector/src/source_collectors/sarif/base.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                        rule=rule["shortDescription"]["text"],
+                        description=rule["fullDescription"]["text"],
+                        url=rule.get("helpUri", ""),
+                    ),
+                )
+        return entities
+
+    @staticmethod
+    def _lookup_result_rule(result: dict, rules: list[dict]) -> dict:
+        """Return the rule corresponding to the result, from either the 'ruleIndex' or 'ruleId' reference."""
```
Comment (fniessink) [inline components/collector/src/source_collectors/sarif/base.py]:
```suggestion
    def _lookup_violated_rule(result: dict, rules: list[dict]) -> dict:
        """Return the rule that was violated, using either the 'ruleIndex' or 'ruleId' reference."""
```

File: components/collector/src/source_collectors/sarif/violations.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+"""SARIF JSON violations collector."""
+from .base import SARIFJSONBase
```
Comment (fniessink) [inline components/collector/src/source_collectors/sarif/violations.py]:
```suggestion
"""SARIF JSON violations collector."""

from .base import SARIFJSONBase
```

File: docs/src/changelog.md [hunk 6]
```diff
@@ -50,6 +50,7 @@ See the [deployment instructions](https://quality-time.readthedocs.io/en/latest/
 
 - When using SonarQube security hotspots as source for the security warnings metric, allow for filtering hotspots by hotspot status ('to review', 'acknowledged', 'fixed', 'safe'). The default value of this parameter is to show hotspots 'to review' and 'acknowledged'. Closes [#5956](https://github.com/ICTU/quality-time/issues/5956).
 - Add axe analysis to Quality-time pipeline. Closes [#5402](https://github.com/ICTU/quality-time/issues/5402).
+- Add support for the SARIF format and `ruleId` lookup, in combination with the Violations metric. Closes [#6314](https://github.com/ICTU/quality-time/issues/6314).
```
Comment (fniessink) [inline docs/src/changelog.md]: The `ruleId` is really an implementation detail of SARIF JSON files. Not sure it's necessary to mention it here. How about: 
```suggestion
- Add support for using the SARIF format as source for the 'violations' metric. This makes it possible to use [Robocop](https://robocop.readthedocs.io/en/stable/) as source for the violations metric. Closes [#6314](https://github.com/ICTU/quality-time/issues/6314).
```

**Review by fniessink — CHANGES_REQUESTED** [review]

File: docs/src/changelog.md (hunk 6 above)
Comment (wkoot) [reply docs/src/changelog.md]: Nice yeah

**Review by fniessink — APPROVED** [review]

**Pull request merged**